File: problems/javascript/07072021-balanced-binary-tree/code.js
// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

const getHeight = (root) => {
  if (!root) return 0

  return 1 + Math.max(getHeight(root.left), getHeight(root.right))
}

/**
 * https://leetcode.com/problems/balanced-binary-tree/
 *
 * @param {TreeNode} root
 * @return {boolean}
 */
const isBalanced = (root) => {
  if (!root) return true

  const leftHeight = getHeight(root.left)
  const rightHeight = getHeight(root.right)

  return (
    Math.abs(leftHeight - rightHeight) <= 1 &&
    isBalanced(root.left) &&
    isBalanced(root.right)
  )
}

File: problems/ruby/04072021-same-tree/code.rb
# Definition for a binary tree node.
class TreeNode
  attr_accessor :val, :left, :right
  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

# https://leetcode.com/problems/same-tree/submissions/
# @param {TreeNode} p
# @param {TreeNode} q
# @return {Boolean}

def is_same_tree(p, q)
  arr_tree_1 = preorder_traversal(p)
  arr_tree_2 = preorder_traversal(q)

  arr_tree_1 == arr_tree_2
end

def preorder_traversal(root)
  return [root] if root.nil?

  [root.val] + preorder_traversal(root.left) + preorder_traversal(root.right)
end
